### DISCUSSION TITLE
Pointer-Interactor

### DISCUSSION THREAD

**Discussion by Ram1508** [microsoft/MixedRealityToolkit-Unity#11636] [Q&A]
@AMollis 

In MRTKv2,OnPointerClicked(MixedRealityPointerEventData eventData) called when pointer is clicked, and one can get the position and rotation of pointer using eventData.

It is replaced by interactors in  mrtk3, but can you tell me how to get the position and rotation of where the interactor is clicked?

**Comment by marlenaklein-msft** [comment]
Hi @Ram1508, thanks for reaching out. 

If you are working with the interactable, there are events [firstSelectEntered](https://docs.unity3d.com/Packages/com.unity.xr.interaction.toolkit@2.0/api/UnityEngine.XR.Interaction.Toolkit.XRBaseInteractable.html#UnityEngine_XR_Interaction_Toolkit_XRBaseInteractable_firstSelectEntered)/[selectEntered](https://docs.unity3d.com/Packages/com.unity.xr.interaction.toolkit@2.0/api/UnityEngine.XR.Interaction.Toolkit.XRBaseInteractable.html#UnityEngine_XR_Interaction_Toolkit_XRBaseInteractable_selectEntered) or [lastSelectExited](https://docs.unity3d.com/Packages/com.unity.xr.interaction.toolkit@2.0/api/UnityEngine.XR.Interaction.Toolkit.XRBaseInteractable.html#UnityEngine_XR_Interaction_Toolkit_XRBaseInteractable_lastSelectExited)/[selectExited](https://docs.unity3d.com/Packages/com.unity.xr.interaction.toolkit@2.0/api/UnityEngine.XR.Interaction.Toolkit.XRBaseInteractable.html#UnityEngine_XR_Interaction_Toolkit_XRBaseInteractable_selectExited) which are fired when an interactor starts or stops selecting an interactable. The args include the interactor responsible. 

If you are working the interactor, it has a [selectEntered](https://docs.unity3d.com/Packages/com.unity.xr.interaction.toolkit@2.0/api/UnityEngine.XR.Interaction.Toolkit.XRBaseInteractor.html#UnityEngine_XR_Interaction_Toolkit_XRBaseInteractor_selectEntered) event, which you can add a listener to. 

Getting the position and rotation may depend on which interactor you are using. The MRTKRayInteractor and GazeInteractor inherit from [XRRayInteractor](https://docs.unity3d.com/Packages/com.unity.xr.interaction.toolkit@2.0/api/UnityEngine.XR.Interaction.Toolkit.XRRayInteractor.html), which provides some functions to get information about the raycast hit. You can use [TryGetHitInfo](https://docs.unity3d.com/Packages/com.unity.xr.interaction.toolkit@2.0/api/UnityEngine.XR.Interaction.Toolkit.XRRayInteractor.html#UnityEngine_XR_Interaction_Toolkit_XRRayInteractor_TryGetHitInfo_UnityEngine_Vector3__UnityEngine_Vector3__System_Int32__System_Boolean__) to get the position of the ray hit and normal of the surface the ray hit or [TryGetCurrent3DRaycastHit](https://docs.unity3d.com/Packages/com.unity.xr.interaction.toolkit@2.0/api/UnityEngine.XR.Interaction.Toolkit.XRRayInteractor.html#UnityEngine_XR_Interaction_Toolkit_XRRayInteractor_TryGetCurrent3DRaycastHit_UnityEngine_RaycastHit__). The XRI docs should provide more info about this.

Hope that this helps!

**Comment by Ram1508** [comment]
Thankyou @marlenaklein-msft  for the reply.
I tried out the above steps you mentioned and below is the screenshot of that,
![image](https://github.com/microsoft/MixedRealityToolkit-Unity/assets/131968892/b3593b4b-d9e1-4b57-af41-2d64533fd15d)

I am not able to receive events(postions) from the TryGetHitInfo()/TryGetCurrent3DRaycastHit().

Prerequistes:

In a scene i have added a gameobject with component 'XRRayInteractor' which has the refrence for rayinteractor[in the above screen shot]

rayinteractor.selectenter.addlistener(OnFirstselectEntered) -> not able to recieve position or normal from onfirstselectentered.

For example I tried to hit a plane with layered as 'SpatialMapping'

**Comment by marlenaklein-msft** [comment]
@Ram1508, ah sorry I misinterpreted this, the selectEntered event is specific to an Interactor selecting an Interactable. If you are working with a plane for spatial mapping I would recommend looking at the MRTK3 [TapToPlace](https://learn.microsoft.com/en-us/windows/mixed-reality/mrtk-unity/mrtk3-spatialmanipulation/packages/spatialmanipulation/solvers/tap-to-place) sample scene and [script](https://github.com/microsoft/MixedRealityToolkit-Unity/blob/mrtk3/com.microsoft.mrtk.spatialmanipulation/Solvers/TapToPlace.cs). I believe this should provide a similar example. Specifically the RegisterPlacementAction and SolverUpdate methods.

**Comment by hemasree99s** [comment]
@marlenaklein-msft, How should we get the Pose where a user has air tapped on the real world without the use of solvers?

**Comment by hemasree99s** [comment]
@AMollis @marlenaklein-msft @mrtk-bld , Any update on the above?

**Comment by AMollis** [comment]
@hemasree99s, you should be able to do this by querying information from Unity's interactors.  

Take a look at a MRTK3's [`TapToPlace`](https://github.com/microsoft/MixedRealityToolkit-Unity/blob/04ad75220add8729fb3c0d4171ad497926a215fa/com.microsoft.mrtk.spatialmanipulation/Solvers/TapToPlace.cs#L494C11-L494C11), and how it queries for the list of interactors.

Then take a look at how [`TapToPlace`](https://github.com/microsoft/MixedRealityToolkit-Unity/blob/04ad75220add8729fb3c0d4171ad497926a215fa/com.microsoft.mrtk.spatialmanipulation/Solvers/TapToPlace.cs#L504) listens for selects (or "clicks").

Using the `SelectEntered` or `SelectEntering` events on the interactor, you will receive a [`SelectEnterEventArgs`](https://docs.unity3d.com/Packages/com.unity.xr.interaction.toolkit@2.0/api/UnityEngine.XR.Interaction.Toolkit.SelectEnterEventArgs.html) object.  This object has the interactable that was selected and interactor that did the selecting.  From this, you can query the attach point via the [`GetAttachPoseOnSelect`](https://docs.unity3d.com/Packages/com.unity.xr.interaction.toolkit@2.0/api/UnityEngine.XR.Interaction.Toolkit.IXRSelectInteractable.html#UnityEngine_XR_Interaction_Toolkit_IXRSelectInteractable_GetAttachPoseOnSelect_UnityEngine_XR_Interaction_Toolkit_IXRSelectInteractor_) method.

For the select events to work on the spatial mesh, you'll have to add an interactable component to your spatial mesh prefab (see your `ARMeshManager` for a reference to your spatial mesh prefab).  If you don't want to add an interactable component to the spatial mesh, you'll then need another way to detect interactor's select actions and execute your own ray cast, much like MRTK3's `TapToPlace` component (see [this](https://github.com/microsoft/MixedRealityToolkit-Unity/blob/04ad75220add8729fb3c0d4171ad497926a215fa/com.microsoft.mrtk.spatialmanipulation/Solvers/TapToPlace.cs#L413C75-L413C75) and [this](https://github.com/microsoft/MixedRealityToolkit-Unity/blob/04ad75220add8729fb3c0d4171ad497926a215fa/com.microsoft.mrtk.spatialmanipulation/Solvers/TapToPlace.cs#L500C67-L500C67)).  Interactors will only interact with interables. So if you want an interactor's intersection with a non-interactable, you'll need to do you own ray casting. 


As you know MRTK3 is based on XRI.  I would also recommend taking a look at Unity's XRI samples given your scenario.

**Reply by hemasree99s** [reply]
@AMollis @marlenaklein-msft 
1.) I tried to add an Interactable to ARMeshManager but subscribed events didn't get fired.
 
2.)When I try to hit a plane with spatial mapping layer, the event subscribed with actionController.selectAction.action.performed (For eg - StopPlacementViaPerformedAction in TapToPlace ) is getting fired. Hence, I need to explicitly do a raycast inorder to find the raycast hit point. 

3.)Anyways to receive the actions subscribed with selectInteractor.selectEntered (For eg- selectInteractor.selectEnteredAddListener(StopPlacementViaSelect) in TapToPlace ) for a non-interactable so that i can get the  attach point via the [GetAttachPoseOnSelect](https://docs.unity3d.com/Packages/com.unity.xr.interaction.toolkit@2.0/api/UnityEngine.XR.Interaction.Toolkit.IXRSelectInteractable.html#UnityEngine_XR_Interaction_Toolkit_IXRSelectInteractable_GetAttachPoseOnSelect_UnityEngine_XR_Interaction_Toolkit_IXRSelectInteractor_) method.

**Reply by hemasree99s** [reply]
@AMollis @marlenaklein-msft 
 Any update on the above request?

**Reply by AMollis** [reply]
@hemasree99s can you please describe your scenario.  What are you trying to do?   

1.  A sample project showing your reproduction would help here.  The physics layer might be filtered out.
2. Yes, that is expected.  This is exactly how the TapToPlace solver works.
3. XRI events only work with interactables.  If you want to work with non-interactables, you need to do your own ray-casting.  I'm curious however, what interaction scenario would require non-interactables?  Whats the reason you can't add an interactable component to the object?

If you have a sample project, I can take a quick look to see what's going wrong.  Do you have something to share?

**Reply by AMollis** [reply]
Also, if you have XRI questions or need help with XRI concepts.  I encourage you to post in the Unity forums as well, https://forum.unity.com/forums/xr-interaction-toolkit-and-input.519/.  There are experts there that might be able to help with your particular scenario.

**Comment by hemasree99s** [comment]
@AMollis, Thanks for the timely help.
